e2e/run_tests.js

```diff
@@ -77,21 +120,128 @@ function writeJsonToFile(jsonObject, filename, dir) {
         catch((err) => console.error(err));
 }
 
+const grepCommand = (word = '') => {
+    // -r, recursive search on subdirectories
+    // -I, ignore binary
+    // -l, only names of files to stdout/return
+    // -w, expression is searched for as a word
+    return `grep -rIlw '${word}' ${TEST_DIR}`;
+};
+
+const grepFiles = (command) => {
+    return shell.exec(command, {silent: true}).stdout.
+        split('\n').
+        filter((f) => f.includes('spec.js'));
+};
+
+function getTestFiles() {
+    const {invert, group, stage} = argv;
+
+    const allFiles = grepFiles(grepCommand());
+
+    const stageFiles = [];
+    if (stage) {
+        const sc = grepCommand(stage.split(',').join('\\|'));
+        stageFiles.push(...grepFiles(sc));
+    }
+    const groupFiles = [];
+    if (group) {
+        const gc = grepCommand(group.split(',').join('\\|'));
+        groupFiles.push(...grepFiles(gc));
+    }
+
+    if (invert) {
+        // Return no test file if no stage and group, but inverted
+        if (!stage && !group) {
+            return [];
+        }
+
+        // Return all excluding stage files
+        if (stage && !group) {
+            return without(allFiles, ...stageFiles);
+        }
+
+        // Return all excluding group files
+        if (!stage && group) {
+            return without(allFiles, ...groupFiles);
+        }
+
+        // Return all excluding group and stage files
+        return without(allFiles, ...intersection(stageFiles, groupFiles));
+    }
+
+    // Return all files if no stage and group flags
+    if (!stage && !group) {
+        return allFiles;
+    }
+
+    // Return stage files if no group flag
+    if (stage && !group) {
+        return stageFiles;
+    }
+
+    // Return group files if no stage flag
+    if (!stage && group) {
+        return groupFiles;
+    }
+
+    // Return files if both in stage and group
+    return intersection(stageFiles, groupFiles);
+}
+
+function getSkippedFiles(initialTestFiles, platform, browser, headless) {
+    const platformFiles = grepFiles(grepCommand(`@${platform}`));
+    const browserFiles = grepFiles(grepCommand(`@${browser}`));
+    const headlessFiles = grepFiles(grepCommand(`@${headless ? 'headless' : 'headed'}`));
```

Comment:
If headless='headed' and some spec files have `Skip: @headed`, wouldn't that skip those files even if the intent was to run headed on those files? Or is `@headed` not allowed in Skip: only `@headless`?

Reply:
I'd say, the intention is not correct. It's allowed to skip both of those.